File: 108.AVL-TreeAlgo.py
# An AVL Tree is a self-balancing Binary Search Tree (BST) where 
# the height difference (balance factor) between the left and right subtrees 
# of every node is at most 1.
# This ensures that search, insert, and delete operations always take O(log n) time.

# Explanation

# In a Binary Search Tree (BST), inserting nodes in sorted order 
# makes the tree skewed and degrades performance to O(n).

# An AVL Tree fixes this by rotating nodes whenever the balance 
# factor becomes less than -1 or greater than +1 after insertion or deletion.

# There are four types of rotations used to restore balance:

# Right Rotation (LL Rotation)

# Left Rotation (RR Rotation)

# Left-Right Rotation (LR Rotation)

# Right-Left Rotation (RL Rotation)

# Each rotation ensures that the height of the tree remains balanced, 
# preserving the logarithmic time complexity for operations.

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 1

class AVLTree:
    def getHeight(self, root):
        if not root:
            return 0
        return root.height
    
    def getBalance(self, root):
        if not root:
            return 0
        return self.getHeight(root.left) - self.getHeight(root.right)
    
    def rightRotate(self, y):
        x = y.left
        T2 = x.right
        x.right = y
        y.left = T2
        y.height = 1 + max(self.getHeight(y.left), self.getHeight(y.right))
        x.height = 1 + max(self.getHeight(x.left), self.getHeight(x.right))
        return x
    
    def leftRotate(self, x):
        y = x.right
        T2 = y.left
        y.left = x
        x.right = T2
        x.height = 1 + max(self.getHeight(x.left), self.getHeight(x.right))
        y.height = 1 + max(self.getHeight(y.left), self.getHeight(y.right))
        return y
    
    def insert(self, root, key):
        if not root:
            return Node(key)
        elif key < root.key:
            root.left = self.insert(root.left, key)
        else:
            root.right = self.insert(root.right, key)

        root.height = 1 + max(self.getHeight(root.left), self.getHeight(root.right))

        balance = self.getBalance(root)

        if balance > 1 and key < root.left.key:
            return self.rightRotate(root)
        
        if balance < -1 and key > root.right.key:
            return self.leftRotate(root)
        
        if balance > 1 and key > root.left.key:
            root.left = self.leftRotate(root.left)
            return self.rightRotate(root)
        
        if balance < -1 and key < root.right.key:
            root.right = self.rightRotate(root.right)
            return self.leftRotate(root)
        
        return root
    
    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        print(root.key, end = " ")
        self.inorder(root.right)

tree = AVLTree()
root = None
nums = [10, 20, 30, 40, 50, 25]

for num in nums:
    root = tree.insert(root, num)

print("Inorder Traversal of the constructed AVL tree is:")
tree.inorder(root)
print()
        

File: 108.AVL-TreeAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;

// Node class representing a single node in the AVL Tree
class Node {
    int key;               // Value stored in the node
    int height;            // Height of the node
    Node left, right;      // References to left and right child nodes

    Node(int d) {          // Constructor to initialize the node
        key = d;
        height = 1;        // Every new node starts with height = 1
    }
}

// Main AVL Tree class
class AVLTree {

    // Helper method to get height of a node
    int height(Node N) {
        if (N == null) return 0;
        return N.height;
    }

    // Helper method to get balance factor of a node
    int getBalance(Node N) {
        if (N == null) return 0;
        return height(N.left) - height(N.right);
    }

    // Right Rotation (LL Rotation)
    Node rightRotate(Node y) {
        Node x = y.left;     // x becomes new root of this subtree
        Node T2 = x.right;   // Temporarily store right subtree of x

        x.right = y;         // Perform rotation
        y.left = T2;

        // Update heights of y and x
        y.height = Math.max(height(y.left), height(y.right)) + 1;
        x.height = Math.max(height(x.left), height(x.right)) + 1;

        return x;            // Return new root after rotation
    }

    // Left Rotation (RR Rotation)
    Node leftRotate(Node x) {
        Node y = x.right;    // y becomes new root of this subtree
        Node T2 = y.left;    // Temporarily store left subtree of y

        y.left = x;          // Perform rotation
        x.right = T2;

        // Update heights of x and y
        x.height = Math.max(height(x.left), height(x.right)) + 1;
        y.height = Math.max(height(y.left), height(y.right)) + 1;

        return y;            // Return new root after rotation
    }

    // Insertion method with balancing
    Node insert(Node node, int key) {
        // 1. Perform normal BST insertion
        if (node == null)
            return (new Node(key));

        if (key < node.key)
            node.left = insert(node.left, key);
        else if (key > node.key)
            node.right = insert(node.right, key);
        else
            return node; // Duplicates not allowed

        // 2. Update the height of this ancestor node
        node.height = 1 + Math.max(height(node.left), height(node.right));

        // 3. Get the balance factor
        int balance = getBalance(node);

        // 4. Check balance and perform rotations if needed

        // Case 1: Left Left (LL)
        if (balance > 1 && key < node.left.key)
            return rightRotate(node);

        // Case 2: Right Right (RR)
        if (balance < -1 && key > node.right.key)
            return leftRotate(node);

        // Case 3: Left Right (LR)
        if (balance > 1 && key > node.left.key) {
            node.left = leftRotate(node.left);
            return rightRotate(node);
        }

        // Case 4: Right Left (RL)
        if (balance < -1 && key < node.right.key) {
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }

        // Return unchanged node if balanced
        return node;
    }

    // Inorder traversal prints sorted keys
    void inorder(Node node) {
        if (node != null) {
            inorder(node.left);
            System.out.print(node.key + " ");
            inorder(node.right);
        }
    }
}

// --------------------------- DRIVER CLASS --------------------------- //
public class Main {
    public static void main(String[] args) throws IOException {

        // Creating object for AVLTree
        AVLTree tree = new AVLTree();
        Node root = null;

        // Insert elements into AVL Tree
        int[] nums = {10, 20, 30, 40, 50, 25};

        // Insert all elements one by one
        for (int num : nums)
            root = tree.insert(root, num);

        // Print inorder traversal (should be sorted)
        System.out.println("Inorder Traversal of the constructed AVL Tree is:");
        tree.inorder(root);
        System.out.println();
    }
}
